app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,264 @@
+        if (!isAllNetworkSupported) {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ACTIVE_NETWORK)
+            serviceHandler?.sendMessage(message)
+        } else {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ALL_NETWORKS)
+            serviceHandler?.sendMessage(message)
+        }
+    }
+
+    override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
+        serviceHandler?.removeMessages(MSG_ADD_ACTIVE_NETWORK, null)
+        serviceHandler?.removeMessages(MSG_ADD_ALL_NETWORKS, null)
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onLinkPropertiesChanged")
+        if (!isAllNetworkSupported) {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ACTIVE_NETWORK)
+            serviceHandler?.sendMessage(message)
+        } else {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ALL_NETWORKS)
+            serviceHandler?.sendMessage(message)
+        }
+    }
+
+    /**
+     * Called when the preference(isAddAllNetwork) is changed.
+     * if the preference is true, add only active network
+     * else, add all available network.
+     */
+    fun onPreferenceChanged(){
+        isAllNetworkSupported = persistentState.isAddAllNetwork
+        serviceHandler?.removeMessages(MSG_ADD_ACTIVE_NETWORK, null)
+        serviceHandler?.removeMessages(MSG_ADD_ALL_NETWORKS, null)
+        if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onPreferenceChanged")
+        if (!isAllNetworkSupported) {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ACTIVE_NETWORK)
+            serviceHandler?.sendMessage(message)
+        } else {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ALL_NETWORKS)
+            serviceHandler?.sendMessage(message)
+        }
+    }
+
+    fun getNetworkList(): MutableSet<Network> {
+        return currentNetworkList
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int): Message {
+        val message = Message()
+        message.what = what
+        return message
+    }
+
+    private fun constructMessage(message: Message?, what: Int): Message {
+        var messageValue = message
+        if (messageValue == null) {
+            messageValue = Message()
+        }
+        messageValue.what = what
+        return messageValue
+    }
+
+    // Handles the network messages from the callback from the connectivity manager
+    private class NetworkRequestHandler(looper: Looper?, val connectionMonitor: ConnectionMonitor) : Handler(looper!!) {
+        override fun handleMessage(msg: Message) {
+            processMessage(msg)
```

Comment:
excessive abstraction. there's no need for a `processMessage` here. Instead, inline `processMessage` body here.

Reply:
Made the processMessage body as inline in handleMessage()